### PR TITLE
Stamp created date

### PR DESCRIPTION
**Pull request by AdeHarrison**
# Motivation and Context
Populate case created date/time in response

# What has changed

# How to test?
1. using a census-rm-case-api image built from [here](https://github.com/ONSdigital/census-rm-case-processor/pull/17)

2. make test

3. All tests should pass

# Links
[Trello](https://trello.com/c/MysxkDnC/776-defect-case-processor-should-stamp-created-date-or-case-api-should-infer-it-from-events)

# Screenshots (if appropriate):

### CODE REVIEW

**Comment by AdeHarrison** [comment]
[DEFECT: Case processor should stamp created date or Case API should infer it from Events](https://trello.com/c/MysxkDnC/776-defect-case-processor-should-stamp-created-date-or-case-api-should-infer-it-from-events)

**Review by NickGrantX — CHANGES_REQUESTED** [review]
Unfortunately, the Swagger spec and the convention for returning dates in JSON is to use ISO format and in the Swagger spec it says it should be Zulu time too (i.e. UTC with "Z" on the end). This is a defect, but this PR perpetuates the defect.

File: src/test/resources/test_response1_with_events.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
   "estabType": "Household",
   "uprn": "123456789012345",
   "caseType": null,
-  "createdDateTime": null,
+  "createdDateTime": "2019-05-28 14:19:00.204",
```
Comment (NickGrantX) [inline src/test/resources/test_response1_with_events.json]: This should be `2019-05-28T14:19:00.204Z`

File: src/test/resources/test_response1_without_events.json [hunk 2]
```diff
@@ -6,7 +6,7 @@
   "estabType": "Household",
   "uprn": "123456789012345",
   "caseType": null,
-  "createdDateTime": null,
+  "createdDateTime": "2019-05-28 14:19:00.204",
```
Comment (NickGrantX) [inline src/test/resources/test_response1_without_events.json]: This should be `2019-05-28T14:19:00.204Z`

File: src/test/resources/test_response2_with_events.json [hunk 3]
```diff
@@ -6,7 +6,7 @@
   "estabType": "Household",
   "uprn": "123456789012345",
   "caseType": null,
-  "createdDateTime": null,
+  "createdDateTime": "2019-05-28 14:19:00.204",
```
Comment (NickGrantX) [inline src/test/resources/test_response2_with_events.json]: This should be `2019-05-28T14:19:00.204Z`

File: src/test/resources/test_response2_without_events.json [hunk 4]
```diff
@@ -6,7 +6,7 @@
   "estabType": "Household",
   "uprn": "123456789012345",
   "caseType": null,
-  "createdDateTime": null,
+  "createdDateTime": "2019-05-28 14:19:00.204",
```
Comment (NickGrantX) [inline src/test/resources/test_response2_without_events.json]: This should be `2019-05-28T14:19:00.204Z`

File: src/test/resources/data-test.sql [hunk 5]
```diff
@@ -7,23 +7,23 @@ INSERT INTO casev2.cases (case_ref, abp_code, address_level, address_line1, addr
...
                           estab_arid, estab_type, lad, latitude, longitude,
                           lsoa, msoa, oa, organisation_name, postcode, rgn, state,
-                          town_name, uprn)
+                          town_name, created_date_time, uprn)
 VALUES (10000000, 'RD06', 'U', 'Flat 56 Francombe House', 'Commercial Road',
         'any addressLine3', 'DDR190314000000195675', 'c0d4f87d-9d19-4393-80c9-9eb94f69c460',
         'DDR190314000000113740', 'Household', 'E06000023','51.4463421', '-2.5924477',
         'E01014540', 'E02003043', 'E00073438', null, 'XX1 0XX', 'E12000009', 'ACTIONABLE',
-        'Windleybury', '123456789012345');
+        'Windleybury', '2019-05-28 14:19:00.204', '123456789012345');
```
Comment (NickGrantX) [inline src/test/resources/data-test.sql]: I suggest inserting dates into the DB using ISO format incl. time zone

File: src/test/resources/data-test.sql [hunk 6]
```diff
@@ -7,23 +7,23 @@ INSERT INTO casev2.cases (case_ref, abp_code, address_level, address_line1, addr
...
                           estab_arid, estab_type, lad, latitude, longitude,
                           lsoa, msoa, oa, organisation_name, postcode, rgn, state,
-                          town_name, uprn)
+                          town_name, created_date_time, uprn)
 VALUES (10000001, 'RD06', 'U', 'First And Second Floor Flat',
         '39 Cranbrook Road', 'any addressLine3', 'DDR190314000000239595', '16d79007-9224-448a-9e59-944d9d153fa1',
         'DDR190314000000060908', 'Household', 'E06000023',
         '51.4721166', '-2.5970579', 'E01014669', 'E02003031', 'E00074083', null, 'XX1 0XX', 'E12000009', 'ACTIONABLE',
-        'Windleybury', '123456789012345');
+        'Windleybury', '2019-05-28 14:19:00.204', '123456789012345');
```
Comment (NickGrantX) [inline src/test/resources/data-test.sql]: I suggest inserting dates into the DB using ISO format incl. time zone

File: src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Case.java [hunk 7]
```diff
@@ -41,8 +41,7 @@
 
   @Column private String addressType;
 
-  // TODO remove @Transient when field in schema
-  @Column @Transient private String createdDateTime;
+  @Column private Date createdDateTime;
```
Comment (LukeLoze1) [inline src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Case.java]: LocalDateTime is usually better than the old Date Type

**Review by LukeLoze1 — CHANGES_REQUESTED** [review]

File: src/test/resources/test_response1_with_events.json (hunk 1 above)
Comment (AdeHarrison) [reply src/test/resources/test_response1_with_events.json]: fixed

File: src/test/resources/data-test.sql (hunk 5 above)
Comment (AdeHarrison) [reply src/test/resources/data-test.sql]: fixed

File: src/test/resources/data-test.sql (hunk 6 above)
Comment (AdeHarrison) [reply src/test/resources/data-test.sql]: fixed

File: src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Case.java (hunk 7 above)
Comment (AdeHarrison) [reply src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Case.java]: fixed

File: src/test/resources/test_response2_without_events.json (hunk 4 above)
Comment (AdeHarrison) [reply src/test/resources/test_response2_without_events.json]: fixed

File: src/test/resources/test_response2_with_events.json (hunk 3 above)
Comment (AdeHarrison) [reply src/test/resources/test_response2_with_events.json]: fixed

File: src/test/resources/test_response1_without_events.json (hunk 2 above)
Comment (AdeHarrison) [reply src/test/resources/test_response1_without_events.json]: fixed

File: src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Event.java [hunk 8]
```diff
@@ -17,7 +17,8 @@
 
   @ManyToOne private UacQidLink uacQidLink;
 
-  @Column private Date eventDate;
+  @Column(columnDefinition = "timestamp with time zone")
+  private Date eventDate;
```
Comment (NickGrantX) [inline src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Event.java]: This type should probably be OffsetDateTime to use the new modern java date types

File: src/main/java/uk/gov/ons/census/caseapisvc/model/dto/CaseContainerDTO.java [hunk 9]
```diff
@@ -23,7 +25,8 @@
   @JsonProperty("caseType")
   private String addressType;
 
-  private String createdDateTime;
+  @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSZ")
+  private Date createdDateTime;
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/caseapisvc/model/dto/CaseContainerDTO.java]: This type should probably be OffsetDateTime to use the new modern java date types

Comment (NickGrantX) [inline src/main/java/uk/gov/ons/census/caseapisvc/model/dto/CaseContainerDTO.java]: This type should probably be OffsetDateTime to use the new modern java date types

File: src/main/java/uk/gov/ons/census/caseapisvc/model/dto/EventDTO.java [hunk 10]
```diff
@@ -14,5 +16,6 @@
   private String eventDescription;
 
   @JsonProperty("createdDateTime")
-  private String eventDate;
+  @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSZ")
+  private Date eventDate;
```
Comment (NickGrantX) [inline src/main/java/uk/gov/ons/census/caseapisvc/model/dto/EventDTO.java]: This type should probably be OffsetDateTime to use the new modern java date types

File: src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Case.java [hunk 11]
```diff
@@ -41,8 +41,8 @@
 
   @Column private String addressType;
 
-  // TODO remove @Transient when field in schema
-  @Column @Transient private String createdDateTime;
+  @Column(columnDefinition = "timestamp with time zone")
+  private Date createdDateTime;
```
Comment (NickGrantX) [inline src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Case.java]: This type should probably be OffsetDateTime to use the new modern java date types

**Review by LukeLoze1 — APPROVED** [review]

File: src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Event.java (hunk 8 above)
Comment (AdeHarrison) [reply src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Event.java]: changed

File: src/main/java/uk/gov/ons/census/caseapisvc/model/dto/CaseContainerDTO.java (hunk 9 above)
Comment (AdeHarrison) [reply src/main/java/uk/gov/ons/census/caseapisvc/model/dto/CaseContainerDTO.java]: changed

File: src/main/java/uk/gov/ons/census/caseapisvc/model/dto/EventDTO.java (hunk 10 above)
Comment (AdeHarrison) [reply src/main/java/uk/gov/ons/census/caseapisvc/model/dto/EventDTO.java]: changed

File: src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Case.java (hunk 11 above)
Comment (AdeHarrison) [reply src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Case.java]: changed

File: src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Event.java [hunk 12]
```diff
@@ -17,7 +17,8 @@
 
   @ManyToOne private UacQidLink uacQidLink;
 
-  @Column private Date eventDate;
+  @Column(columnDefinition = "timestamp with time zone")
```
Comment (NickGrantX) [inline src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Event.java:20]: don't need this `columnDefinition` anymore I think

**Review by NickGrantX — CHANGES_REQUESTED** [review]
Just one rogue `columnDefinition` left in

File: src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Event.java (hunk 12 above)
Comment (AdeHarrison) [reply src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Event.java:20]: fixed

File: src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Event.java [hunk 13]
```diff
@@ -17,7 +17,7 @@
 
   @ManyToOne private UacQidLink uacQidLink;
 
-  @Column private Date eventDate;
+  private OffsetDateTime eventDate;
```
Comment (NickGrantX) [inline src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Event.java:21]: This needs the `@Column` annotation

**Review by NickGrantX — CHANGES_REQUESTED** [review]
Just missing the `@Column` annotation on the `eventDate` member of the `Event` entity

File: src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Event.java (hunk 12 above)
Comment (NickGrantX) [reply src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Event.java:20]: it does need `@Column` but not the `columnDefinition`

File: src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Event.java (hunk 13 above)
Comment (AdeHarrison) [reply src/main/java/uk/gov/ons/census/caseapisvc/model/entity/Event.java:21]: done

**Review by NickGrantX — APPROVED** [review]
LGTM 👍

**Pull request merged**